### ISSUE TITLE
Problems configuring network interfaces on windows

### ISSUE THREAD

**Issue by carcus88** [Parallels/vagrant-parallels#170]
I have a windows box that I'm trying to provision and cannot set the mac address. I'm not sure if its a box issue or a vagrant/parallels issue. Basically pulling in the following box

https://vagrantcloud.com/scorebig/boxes/windows-2012R2-SC

and trying to provision the box's network using using a simple vagrant file which includes the following 

```
config.vm.guest = :windows
config.vm.network "private_network", ip: "192.168.56.133"
```
 
results in the following warnings when doing vagrant up --debug

```
INFO interface: info: Configuring and enabling network interfaces...
 INFO interface: info: ==> default: Configuring and enabling network interfaces...
DEBUG winrmcommunicator: Checking whether WinRM is ready...
DEBUG winrmshell: powershell executing:
hostname
DEBUG winrmshell: Exit status: 0
 INFO winrmcommunicator: WinRM is ready!
DEBUG guest: Searching for cap: configure_networks
DEBUG guest: Checking in: windows
DEBUG guest: Found cap: configure_networks in windows
 INFO guest: Execute capability: configure_networks [#<Vagrant::Machine: default (VagrantPlugins::Parallels::Provider)>, [{:type=>:static, :adapter_ip=>"192.168.56.1", :ip=>"192.168.56.133", :netmask=>"255.255.255.0", :auto_config=>true, :interface=>1}, {:type=>:static, :adapter_ip=>"10.37.129.1", :ip=>"10.37.129.150", :netmask=>"255.255.255.0", :auto_config=>true, :interface=>2}]] (windows)
DEBUG configure_networks: networks: [{:type=>:static, :adapter_ip=>"192.168.56.1", :ip=>"192.168.56.133", :netmask=>"255.255.255.0", :auto_config=>true, :interface=>1}, {:type=>:static, :adapter_ip=>"10.37.129.1", :ip=>"10.37.129.150", :netmask=>"255.255.255.0", :auto_config=>true, :interface=>2}]
DEBUG winrmshell: initializing WinRMShell
 INFO subprocess: Starting process: ["/usr/bin/prlctl", "list", "24dd2252-aa00-46e7-b6a6-116a3707dcca", "--no-header", "-o", "mac"]
DEBUG subprocess: Command not in installer, not touching env vars.
DEBUG subprocess: Selecting on IO
DEBUG subprocess: stdout: 00:1C:42:96:FC:A3 00:1C:42:8F:86:6B 00:1C:42:84:DF:8B 
DEBUG subprocess: Waiting for process to exit. Remaining to timeout: 31999
DEBUG subprocess: Exit status: 0
DEBUG configure_networks: mac addresses: {"001C4296FCA3"=>"0", "001C428F866B"=>"1", "001C4284DF8B"=>"2"}
DEBUG winrmshell: querying WSMan version
DEBUG winrmshell: powershell executing:
((test-wsman).productversion.split(" ") | select -last 1).split("\.")[0]
DEBUG winrmshell: Exit status: 0
DEBUG winrmshell: wsman version: 3

DEBUG winrmshell: powershell executing:
$adapters = get-ciminstance win32_networkadapter -filter "macaddress is not null" 
$processed = @()
foreach ($adapter in $adapters) {
  $Processed += new-object PSObject -Property @{
    mac_address = $adapter.macaddress
    net_connection_id = $adapter.netconnectionid
    interface_index = $adapter.interfaceindex
    index = $adapter.index
  }
} 
convertto-json -inputobject $processed

DEBUG winrmshell: Exit status: 0
DEBUG winrmshell: [{:index=>10, :mac_address=>"00:1C:42:96:FC:A3", :net_connection_id=>"Ethernet", :interface_index=>12}, {:index=>13, :mac_address=>"00:1C:42:84:DF:8B", :net_connection_id=>"Ethernet 2", :interface_index=>15}, {:index=>14, :mac_address=>"00:1C:42:8F:86:6B", :net_connection_id=>"Ethernet 3", :interface_index=>16}]
DEBUG configure_networks: nic: {:index=>10, :mac_address=>"00:1C:42:96:FC:A3", :net_connection_id=>"Ethernet", :interface_index=>12}
DEBUG configure_networks: nic: {:index=>13, :mac_address=>"00:1C:42:84:DF:8B", :net_connection_id=>"Ethernet 2", :interface_index=>15}
DEBUG configure_networks: nic: {:index=>14, :mac_address=>"00:1C:42:8F:86:6B", :net_connection_id=>"Ethernet 3", :interface_index=>16}
DEBUG configure_networks: vm_interface_map: {"0"=>{:net_connection_id=>"Ethernet", :mac_address=>"001C4296FCA3", :interface_index=>12, :index=>10}, "2"=>{:net_connection_id=>"Ethernet 2", :mac_address=>"001C4284DF8B", :interface_index=>15, :index=>13}, "1"=>{:net_connection_id=>"Ethernet 3", :mac_address=>"001C428F866B", :interface_index=>16, :index=>14}}
 WARN configure_networks: Could not find interface for network {:type=>:static, :adapter_ip=>"192.168.56.1", :ip=>"192.168.56.133", :netmask=>"255.255.255.0", :auto_config=>true, :interface=>1}
 WARN configure_networks: Could not find interface for network {:type=>:static, :adapter_ip=>"10.37.129.1", :ip=>"10.37.129.150", :netmask=>"255.255.255.0", :auto_config=>true, :interface=>2}
 INFO warden: Calling OUT action: #<VagrantPlugins::Parallels::Action::Network:0x000001033158b8>
```

It seems that the windows box lists the interfaces with an interface index and that these must somehow match the vagrant networking interface indexes is what causes the failure. I'm not sure how windows assigns the interface index or if it can be specified via vagrant in order to match or if there is another method for doing this.

I had reported the same issue to vagrant a while back but got nowhere with it and was told to report it here instead. 

https://github.com/mitchellh/vagrant/issues/4581

**Comment by legal90** [comment]
@carcus88 Thank you for reporting and sorry for a delay.
This issue has been just fixed in [`v1.3.8`](https://github.com/Parallels/vagrant-parallels/releases/tag/v1.3.8). Please, update

**Comment by carcus88** [comment]
Thank you @legal90 !  Was this only an issue for windows guests? It looks like the change affects the MAC address hash for all guests. Would you mind explaining what the issue was?

**Comment by legal90** [comment]
@carcus88 
https://github.com/mitchellh/vagrant/commit/d600e83e6d5f698a389dbfe6516736cde157bfb1 - this commit added a new entity to Vagrant: the provider capability named [`nic_mac_addresses`](https://github.com/mitchellh/vagrant/blob/d600e83e6d5f698a389dbfe6516736cde157bfb1/plugins/providers/virtualbox/cap.rb#L22-L24). It is assumed that this capability will be implemented in all providers.

I've been watching these changes in mainstream Vagrant, so I've implemented this functionality [in the Parallels provider too](https://github.com/Parallels/vagrant-parallels/blob/master/lib/vagrant-parallels/cap/nic_mac_addresses.rb).
But I've missed one little thing: initially, it was assumed that this method will return a hash with integer keys, e.q. like `{0 => '00:1c:42:00:00:09', 1 => 00:1c:42:00:00:08, ...}`:
https://github.com/mitchellh/vagrant/blob/fa38b59049553d26e8f4903bfdaea9fd0478a46a/plugins/providers/virtualbox/driver/version_4_2.rb#L377-L388

But in our case there was a string keys: `{'0' => '00:1c:42:00:00:09', '1' => 00:1c:42:00:00:08, ...}`. So that, this line in windows net configuration was always getting `nil` while trying to fetch an adapter by the integer index: [plugins/guests/windows/cap/configure_networks.rb#L26](https://github.com/mitchellh/vagrant/blob/bb052366f7f9616666299a97edcceadae007394a/plugins/guests/windows/cap/configure_networks.rb#L26)

*Why only windows guests are affected?*
Provider capability `nic_mac_addresses` is not guest specific and could be used wherever in Vagrant code. But in fact it is used only for network configuration for Windows guests. Just try to search "nic_mac_addresses" in the vagrant repository to make sure:
https://github.com/mitchellh/vagrant/search?utf8=%E2%9C%93&q=nic_mac_addresses

**Comment by legal90** [comment]
I've just found that an issue is not resolved completely. Keys are integers now, however Vagrant expects that NIC will be numerated starting from 1 instead of 0, as it is designed in Virtualbox:

So, provider capability `nic_mac_addresses` should return a hash with keys `1, 2, 3, ...`. 
I'll try to fix it soon

**Comment by legal90** [comment]
Fixed in v1.3.9
